### PR TITLE
Check php compatibility of app store app releases

### PR DESCRIPTION
**Pull request by ChristophWurst**
Apps might increase the minimum php version requirement, in which case
an update could break the app or even the whole instance. We must not
install those releases, or better, don't even show them for
update/installation. This extends the app fetcher code to filter out the
releases that are not installable.

The filter respects minimum and maximum requirements. E.g. apps that are
still only released for php7.3 won't show up for php7.4 instances. This
behavior is new but if an app lists an explicit version requirement,
then we ought to repect that.

Apps to test this with are `twofactor_totp`, `twofactor_u2f` and `ocr` as those three only allow php7.3 if we go by the info.xml.

S/o to @nickvergessen for the very useful pointers.

### CODE REVIEW

File: lib/private/App/AppStore/Fetcher/AppFetcher.php [hunk 1]
```diff
@@ -101,14 +102,33 @@ protected function fetch($ETag, $content, $allowUnstable = false) {
...
+						$maxServerVersion = $serverVersion->getMaximumVersion();
+						$minFulfilled = $this->compareVersion->isCompatible($ncVersion, $minServerVersion, '>=');
+						$maxFulfilled = $maxServerVersion !== '' &&
+							$this->compareVersion->isCompatible($ncVersion, $maxServerVersion, '<=');
+						$isPhpCompatible = true;
+						if (($release['rawPhpVersionSpec'] ?? '*') !== '*') {
+							$phpVersion = $versionParser->getVersion($release['rawPhpVersionSpec']);
+							$minPhpVersion = $phpVersion->getMinimumVersion();
+							$maxPhpVersion = $phpVersion->getMaximumVersion();
+							$currentPhpVersion = phpversion();
```
Comment (kesselb) [inline lib/private/App/AppStore/Fetcher/AppFetcher.php:117]:
```suggestion
```

File: lib/private/App/AppStore/Fetcher/AppFetcher.php [hunk 2]
```diff
@@ -101,14 +102,33 @@ protected function fetch($ETag, $content, $allowUnstable = false) {
...
+							$minPhpFulfilled = $minPhpVersion === '' || $this->compareVersion->isCompatible(
+									$currentPhpVersion,
```
Comment (kesselb) [inline lib/private/App/AppStore/Fetcher/AppFetcher.php:119]:
```suggestion
									PHP_VERSION,
```

File: lib/private/App/AppStore/Fetcher/AppFetcher.php [hunk 3]
```diff
@@ -101,14 +102,33 @@ protected function fetch($ETag, $content, $allowUnstable = false) {
...
+									$minPhpVersion,
+									'>='
+								);
+							$maxPhpFulfilled = $maxPhpVersion === '' || $this->compareVersion->isCompatible(
+									$currentPhpVersion,
```
Comment (kesselb) [inline lib/private/App/AppStore/Fetcher/AppFetcher.php:124]:
```suggestion
									PHP_VERSION,
```

**Comment by StCyr** [comment]
> We must not
> install those releases, or better, don't even show them for
> update/installation. This extends the app fetcher code to filter out the
> releases that are not installable.

I don't agree with you that not showing them is better.

For example, I've already experienced -and also seen someone reports- an issue where NC wasn't showing me a new realease because I was running an old version of php. It tooks me weeks -and some contacts with NC's community- to realize that I wasn't shown the new version because of this.

So, I believe it's much better to display a message to the user that a new version is available but can't be installed because of missing requirements

**Comment by ChristophWurst** [comment]
Fair enough, that would be ideal indeed.

But UI wise that is a big change and I don't currently have the time to restructure the apps management for that change.

This PR fixes the current bug that Nextcloud blindly installs incompatible apps. Preventing this is already a huge step forward. Showing the admin that they could update the app even further if certain dependencies are met (Nextcloud server, php, any other dependency that we allow specifying) is the next logical step.

**Comment by StCyr** [comment]
While looking at this, I've found:

https://github.com/nextcloud/server/blob/785029d69332d83773a19c75077f25779fabd55a/lib/private/Installer.php#L398

'Seems related to this PR too...

**Review by juliusknorr — APPROVED** [review]

**Comment by ChristophWurst** [comment]
🏓 please review :)

**Review by rullzer — APPROVED** [review]

**Pull request merged**